openmc/data/resonance_covariance.py

```diff
@@ -0,0 +1,822 @@
+        """Produce a subset of resonance parameters and the corresponding
+        covariance matrix to an IncidentNeutron object.
+
+        Parameters
+        ----------
+        parameter_str : str
+            parameter to be discriminated
+            (i.e. 'energy', 'captureWidth', 'fissionWidthA'...)
+        bounds : np.array
+            [low numerical bound, high numerical bound]
+
+        Returns
+        -------
+        res_cov_range : openmc.data.ResonanceCovarianceRange
+            ResonanceCovarianceRange object that contains a subset of the
+            covariance matrix (upper triangular) as well as a subset parameters
+            within self.file2params
+
+        """
+        # Copy range and prevent change of original
+        res_cov_range = copy.deepcopy(self)
+
+        parameters = self.file2res.parameters
+        cov = res_cov_range.covariance
+        mpar = res_cov_range.mpar
+        # Create mask
+        mask1 = parameters[parameter_str] >= bounds[0]
+        mask2 = parameters[parameter_str] <= bounds[1]
+        mask = mask1 & mask2
+        res_cov_range.parameters = parameters[mask]
+        indices = res_cov_range.parameters.index.values
+        # Build subset of covariance
+        sub_cov_dim = len(indices)*mpar
+        cov_subset_vals = []
+        for index1 in indices:
+            for i in range(mpar):
+                for index2 in indices:
+                    for j in range(mpar):
+                        if index2*mpar+j >= index1*mpar+i:
+                            cov_subset_vals.append(cov[index1*mpar+i,
+                                                   index2*mpar+j])
+
+        cov_subset = np.zeros([sub_cov_dim, sub_cov_dim])
+        tri_indices = np.triu_indices(sub_cov_dim)
+        cov_subset[tri_indices] = cov_subset_vals
+
+        res_cov_range.file2res.parameters = parameters[mask]
+        res_cov_range.covariance = cov_subset
+        # Set _prepared to False to ensure parameter subset
+        # used during construction routine
+        res_cov_range.file2res._prepared = False
+
+        return res_cov_range
+
+    def sample_resonance_parameters(self, n_samples):
+        """Sample resonance parameters based on the covariances provided
+        within an ENDF evaluation.
+
+        Parameters
+        ----------
+        n_samples : int
+            The number of samples to produce
+
+        Returns
+        -------
+        samples : list of openmc.data.ResonanceCovarianceRange objects
+            List of samples size `n_samples`
+
+        """
+        warn_str = 'Sampling routine does not guarantee positive values for '\
+                   'parameters. This can lead to undefined behavior in the '\
+                   'reconstruction routine.'
+        warnings.warn(warn_str)
+        parameters = self.parameters
+        cov = self.covariance
+        # Copy ResonanceRange object
+        res_range = copy.copy(self.file2res)
```

Comment:
Right now, you're creating a single copy of the resonance range. If multiple samples are requested, this method actually returns a list with the same object repeated N times (this is evident in the Jupyter notebook where you can see that two different samples have the same parameters). You should be creating N copies of the original object and then setting the parameters appropriately for each one.

Reply:
Saw similar code being used for each formalism/mpar so thought I could pull it out front but guess copy.copy() and .append() behaved differently than I thought. Should be better now.